src/provider/provider_os_memory_windows.c

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <Windows.h>
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <umf/providers/provider_os_memory.h>
+
+int os_translate_mem_protection_flags(unsigned protection) {
+    switch (protection) {
+    case UMF_PROTECTION_NONE:
+        return PAGE_NOACCESS;
+    case UMF_PROTECTION_EXEC:
+        return PAGE_EXECUTE;
+    case (UMF_PROTECTION_EXEC | UMF_PROTECTION_READ):
+        return PAGE_EXECUTE_READ;
+    case (UMF_PROTECTION_EXEC | UMF_PROTECTION_READ | UMF_PROTECTION_WRITE):
+        return PAGE_EXECUTE_READWRITE;
+    case (UMF_PROTECTION_EXEC | UMF_PROTECTION_WRITE):
+        return PAGE_EXECUTE_WRITECOPY;
+    case UMF_PROTECTION_READ:
+        return PAGE_READONLY;
+    case (UMF_PROTECTION_READ | UMF_PROTECTION_WRITE):
+        return PAGE_READWRITE;
+    case UMF_PROTECTION_WRITE:
+        return PAGE_WRITECOPY;
+    }
+    fprintf(stderr,
+            "os_translate_mem_protection_flags(): unsupported protection flag: "
+            "%u\n",
+            protection);
+    assert(0);
+    return -1;
+}
+
+void *os_mmap(void *hint_addr, size_t length, int prot) {
+    return VirtualAlloc(hint_addr, length, MEM_RESERVE | MEM_COMMIT, prot);
+}
+
+int os_munmap(void *addr, size_t length) {
+    // If VirtualFree() succeeds, the return value is nonzero.
+    // If VirtualFree() fails, the return value is 0 (zero).
+    (void)length; // unused
+    return (VirtualFree(addr, 0, MEM_RELEASE) == 0);
+}
+
+int os_purge(void *addr, size_t length, int advice) {
+    fprintf(stderr, "os_purge() NOT IMPLEMENTED\n");
```

Comment:
I think this is what the MEM_RELEASE flag does

Reply:
`MEM_RELEASE` does unmapping (=`munmap()`)

Reply:
IMHO `umfProviderPurge()` should **NOT** do exactly the same what `umfProviderFree()` does (e.g. `VirtualFree(addr, 0, MEM_RELEASE)`), otherwise `umfProviderFree()` would always fail when called after `umfProviderPurge()`

Reply:
But you can't call the free after calling the purge - please refer to e.g. jemalloc documentation. free() is weaker than purge(). free() means that the pointer shouldn't be used to access a memory but there are no requirements on what to do with this memory (you could either destroy == unmap it or return to the pool) while the purge() means that you free this memory AND destroys it which returns it to the OS or device. You could do this immediately (force version) or later (lazy version) as it potentially is costly. @igchor please correct me if I'm wrong

Reply:
but jemalloc calls in a sequence its hooks:
1) `arena_extent_purge_lazy()` which calls `umfMemoryProviderPurgeLazy()` then:
2) `arena_extent_purge_forced()` which calls `umfMemoryProviderPurgeForce()` and finally:
3) `arena_extent_destroy()` which calls `umfMemoryProviderFree()` and finally:

Reply:
> But you can't call the free after calling the purge - please refer to e.g. jemalloc documentation. free() is weaker than purge(). 

AFAIK it is the opposite: you can't call purge() after calling free() - free() unmaps the memory and you cannot do anything else with it later. On Linux `purge_lazy()` is implemented as `madvise(MADV_FREE)` and `purge_force()` is implemented as `madvise(MADV_DONTNEED)`, but `madvise()` can be called only on the mapped region of memory, so if you free (munmap()) it, you cannot do anything else more with it.

Reply:
ok I'm resolving this as we agreed that we'll fix the provider API in the separate PR